reactor-core/src/main/java/reactor/core/publisher/MonoCacheInvalidateIf.java

```diff
@@ -0,0 +1,389 @@
+			return null;
+		}
+
+		@Override
+		public void clear() {
+			//NO-OP
+		}
+	};
+
+	final Predicate<? super T> shouldInvalidatePredicate;
+
+	volatile     State<T>                                                    state;
+	@SuppressWarnings("rawtypes")
+	static final AtomicReferenceFieldUpdater<MonoCacheInvalidateIf, State> STATE =
+			AtomicReferenceFieldUpdater.newUpdater(MonoCacheInvalidateIf.class, State.class, "state");
+
+	MonoCacheInvalidateIf(Mono<T> source, Predicate<? super T> invalidationPredicate) {
+		super(source);
+		this.shouldInvalidatePredicate = Objects.requireNonNull(invalidationPredicate, "invalidationPredicate");
+		@SuppressWarnings("unchecked")
+		State<T> state = (State<T>) EMPTY_STATE;
+		this.state = state;
+	}
+
+	@Override
+	public CoreSubscriber<? super T> subscribeOrReturn(CoreSubscriber<? super T> actual) {
+		CacheMonoSubscriber<T> inner = new CacheMonoSubscriber<>(actual);
+		//important: onSubscribe should be deferred until we're sure we're in the Coordinator case OR the cached value passes the predicate
+		for(;;) {
+			State<T> state = this.state;
+			if (state == EMPTY_STATE || state instanceof CoordinatorSubscriber) {
+				boolean connectToUpstream = false;
+				CoordinatorSubscriber<T> coordinator;
+				if (state == EMPTY_STATE) {
+					coordinator = new CoordinatorSubscriber<>(this, this.source);
+					if (!STATE.compareAndSet(this, EMPTY_STATE, coordinator)) {
+						continue;
+					}
+					connectToUpstream = true;
+				}
+				else {
+					coordinator = (CoordinatorSubscriber<T>) state;
+				}
+
+				if (coordinator.add(inner)) {
+					if (inner.isCancelled()) {
+						coordinator.remove(inner);
+					}
+					else {
+						inner.coordinator = coordinator;
+						actual.onSubscribe(inner);
+					}
+
+					if (connectToUpstream) {
+						coordinator.delayedSubscribe();
+					}
+					return null;
+				}
+			}
+			else {
+				//state is an actual signal, cached
+				T cached = state.get();
+				try {
+					boolean invalidated = this.shouldInvalidatePredicate.test(cached);
+					if (invalidated) {
+						if (STATE.compareAndSet(this, state, EMPTY_STATE)) {
+							Operators.onDiscard(cached, actual.currentContext());
+						}
+						//we CAS but even if it fails we want to loop back
+						continue;
+					}
+				}
+				catch (Throwable error) {
+					if (STATE.compareAndSet(this, state, EMPTY_STATE)) {
+						Operators.onDiscard(cached, actual.currentContext());
+						Operators.error(actual, error);
+						return null;
+					}
+				}
+
+
+				actual.onSubscribe(inner);
+				inner.complete(state.get());
+				return null;
+			}
+		}
+	}
+
+	@Override
+	public Object scanUnsafe(Attr key) {
+		if (key == Attr.RUN_STYLE) return Attr.RunStyle.SYNC;
+		return super.scanUnsafe(key);
+	}
+
+	static final class CoordinatorSubscriber<T> implements InnerConsumer<T>, State<T> {
+
+		final MonoCacheInvalidateIf<T> main;
+		final Mono<? extends T> source;
+
+		volatile boolean gotOnNext = false;
```

Comment:
this doesn't need to be `volatile`, could be called `done` like in other operators